tests/test_single_controller.py

```diff
@@ -0,0 +1,308 @@
+# Copyright 2024 MosaicML ComposeRL authors
+# SPDX-License-Identifier: Apache-2.0
+
+import logging
+import os
+import pathlib
+from datetime import timedelta
+from typing import Optional
+
+import pytest
+import ray
+import torch
+import torch.distributed as dist
+from transformers import (
+    AutoModelForCausalLM,
+    PreTrainedModel,
+    PreTrainedTokenizerBase,
+)
+
+from compose_rl.algorithms.online.generation_utils import (
+    create_vllm_engines,
+    init_process_group,
+)
+from compose_rl.utils.ray_utils import (
+    get_free_port,
+    get_node_ip,
+    is_cuda_visible_devices_set,
+    start_ray_server,
+)
+from tests.common import world_size
+
+# Set up logging
+logger = logging.getLogger(__name__)
+
+
+@ray.remote(num_gpus=1)
+class DistributedGPUActor:
+
+    def __init__(
+        self,
+        rank: int,
+        world_size: int,
+        master_addr: Optional[str] = None,
+        master_port: Optional[int] = None,
+    ):
+        """Initialize the distributed GPU actor.
+
+        Args:
+            rank: The rank of this process in the distributed group
+            world_size: Total number of processes in the distributed group
+            master_addr: Master node address. If None, will allocate dynamically for rank 0
+            master_port: Master node port. If None, will allocate dynamically for rank 0
+        """
+        self.rank = rank
+        self.world_size = world_size
+        self.master_addr = master_addr
+        self.master_port = master_port
+
+        # Set up basic environment variables
+        os.environ['WORLD_SIZE'] = str(world_size)
+        os.environ['RANK'] = str(rank)
+
+        # Set LOCAL_RANK based on Ray GPU allocation
+        os.environ['LOCAL_RANK'] = '0' if is_cuda_visible_devices_set(
```

Comment:
Seems that setting all nodes' `LOCAL_RANK=0` would lead to 2D parallelism concerns down the line, right?